documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,886 @@
+>
+> DASH-SAI pipeline is a logical concept. It doesn't have to be 1 to 1 mapping to a physical ASIC pipeline. The underlying implementation can be as simple as a metadata field update with a specific value, when the entry getting matched, e.g., `ENI = Inner Source/Destination MAC`.
+
+### 5.5. Packet Decap
+
+If a pipeline is found, before processing the packets, all outer encaps will be decap'ed, and with key information saved in metadata bus, such as encap type, source IP and VNI / GRE Key, exposing the inner most packet going through the pipeline. This simplifies the flow matching logic and also allow us to create the reverse flow properly.
+
+#### 5.5.1. Multi-layer encap handling
+
+The incoming packet to DASH pipeline could have multiple layers of encap, however, the ASIC capacity is usually limited. The more encaps we handle in parser and decaps, the more the latency will be. Furthermore, using packet recirculation is also not ideal, because it can greatly reduces the throughput and increases the latency.
+
+The limit of having 2 layers of encap at maximum is trying to help this problem and avoid the pipeline having unexpected latency.
+
+#### 5.5.2. Stateless decap vs stateful decap
+
+During the direction lookup stage, all other encaps will be examined, such as VNI lookup. For each VNI or GRE key, we can specify whether it is stateless or stateful.
+
+Although the encap information will still be saved in the metadata bus, however, for stateless decap, we will not create the encap for the reverse flow. Otherwise (for the stateful decap), we will create the encap for the reverse flow by reversing the source and destination information.
+
+```json
+{
+    "DASH_SAI_VNI_TABLE|12345": { "direction": "outbound", "stateless": false },
+    "DASH_SAI_VNI_TABLE|12346": { "direction": "outbound", "stateless": true }
+}
+```
+
+#### 5.5.3. Encap fields handling
+
+##### 5.5.3.1. Handling DSCP
+
+DASH pipeline provides 2 modes for handling the DSCP: "Preserve model" and "Pipe model".
+
+Preserve model works very similar to `SAI_TUNNEL_DSCP_MODE_UNIFORM_MODEL` with a slight difference:
+
+- The DSCP value of outer most encap header will be preserved.
+- When any encap being added, the preserved DSCP value will be copied to the new encap header.
+- The DSCP value of overlay packet (customer packet) shall never be modified or preserved and copied to the outer header.
+
+On the other hand, pipe model works very similar to `SAI_TUNNEL_DSCP_MODE_PIPE_MODEL`:
+
+- The DSCP value of existing encaps will not be preserved.
+- When any encap being added, the DASH pipeline will specify the DSCP value, and all routing action will take this value and copy it to the encaps.
+
+If no encaps are added, say, for traffic sending to Internet, the DSCP value of overlay packet will be exposed as it is, even-if original packet arrives with a non-zero value.
+
+This gives the cloud infra full control over the DSCP value and avoid the customer packet spoofing the fields.
```

Comment:
```suggestion
This gives the cloud infra full control over the DSCP value and will prevent the customer from packet spoofing the fields.
```

Reply:
i took your advice and updated this part slightly different. hope it makes sense.